launcher/main.go

```diff
@@ -139,57 +172,343 @@ func readJsonArgs(filename string) (args []string) {
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		if failed {
+			dbusSyncRead.Close()
+			dbusSyncWrite.Close()
+		}
+	}()
+	cmd.ExtraFiles = []*os.File{dbusSyncWrite}
+
+	dbus.Cmd = cmd
+	dbus.SyncRead = dbusSyncRead
+
+	if err := cmd.Start(); err != nil {
+		panic(err)
+	}
+	defer func() {
+		if failed {
+			dbus.Close()
+		}
+	}()
+
+	if err := dbusSyncWrite.Close(); err != nil {
+		panic(err)
+	}
+
+	failed = false
+	return
+}
+
+func (dbus *Dbus) WaitUntilStartup() {
+	if _, err := dbus.SyncRead.Read([]byte{'x'}); err != nil {
+		panic(err)
+	}
+}
+
+func (dbus *Dbus) Close() {
+	dbus.SyncRead.Close()
+	dbus.Cmd.Wait()
+}
+
+type BwrapInfo struct {
+	ChildPid int `json:"child-pid"`
+}
+
+type Bwrap struct {
+	Cmd        *exec.Cmd
+	InfoRead   *os.File
+	BlockWrite *os.File
+	Info       BwrapInfo
+}
+
+func StartBwrap(conf Config) (bwrap Bwrap) {
+	failed := true
+
+	bwrapArgs := append([]string{"--info-fd", "3", "--block-fd", "4"}, conf.BwrapArgs...)
+	bwrapArgs = append(bwrapArgs, "--")
+	bwrapArgs = append(bwrapArgs, conf.AppExe)
+	bwrapArgs = append(bwrapArgs, conf.AppArgs...)
+
+	cmd := exec.Command(conf.BwrapExe, bwrapArgs...)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	bwrapInfoRead, bwrapInfoWrite, err := os.Pipe()
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		if failed {
+			bwrapInfoRead.Close()
+			bwrapInfoWrite.Close()
+		}
+	}()
+	bwrapBlockRead, bwrapBlockWrite, err := os.Pipe()
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		if failed {
+			bwrapBlockRead.Close()
+			bwrapBlockWrite.Close()
+		}
+	}()
+	cmd.ExtraFiles = []*os.File{bwrapInfoWrite, bwrapBlockRead}
+
+	bwrap.Cmd = cmd
+	bwrap.InfoRead = bwrapInfoRead
+	bwrap.BlockWrite = bwrapBlockWrite
+
+	if err := cmd.Start(); err != nil {
+		panic(err)
+	}
+	defer func() {
+		if failed {
+			bwrap.Close()
+		}
+	}()
+
+	if err := bwrapInfoWrite.Close(); err != nil {
+		panic(err)
+	}
+	if err := bwrapBlockRead.Close(); err != nil {
+		panic(err)
+	}
+
+	failed = false
+	return
+}
+
+func (bwrap *Bwrap) WaitUntilSandboxReady() (bwrapInfo BwrapInfo) {
+	if bytes, err := ioutil.ReadAll(bwrap.InfoRead); err == nil {
+		if err := json.Unmarshal(bytes, &bwrapInfo); err != nil {
+			panic(err)
+		}
+		if bwrapInfo.ChildPid <= 0 {
+			panic("Unexpected child PID")
+		}
+		bwrap.Info = bwrapInfo
+	} else {
+		panic(err)
+	}
+	if err := bwrap.InfoRead.Close(); err != nil {
+		panic(err)
+	}
+
+	return
+}
+
+func (bwrap *Bwrap) NotifySandboxFinished() {
+	if _, err := bwrap.BlockWrite.Write([]byte{'x'}); err != nil {
+		panic(err)
+	}
+	if err := bwrap.BlockWrite.Close(); err != nil {
+		panic(err)
+	}
+}
+
+func (bwrap *Bwrap) WaitUntilParentExit() error {
+	return bwrap.Cmd.Wait()
+}
+
+func (bwrap *Bwrap) WaitUntilChildExit() {
+	bwrapChild, err := os.FindProcess(bwrap.Info.ChildPid)
+	if err != nil {
+		panic(err)
+	}
+	if _, err := bwrapChild.Wait(); err != nil {
```

Comment:
There seems to be a race condition here. I guess you'd need to check for `ECHILD`.
```console
$ nix build .#busybox
$ while ./result/bin/busybox echo ok; do :; done
ok
ok
ok
(...)
panic: waitid: no child processes

goroutine 1 [running]:
main.(*Bwrap).WaitUntilChildExit(0xc000170180?)
	github.com/nixpak/nixpak/launcher/main.go:386 +0x45
main.run()
	github.com/nixpak/nixpak/launcher/main.go:501 +0x3a5
main.main()
	github.com/nixpak/nixpak/launcher/main.go:507 +0x13
```

Reply:
Thanks for the re-producer!
Our new child reaper might catch the child exit before we explicitly wait for it. Since we do not need the exit code, I think it is fine to just let the child reaper catch it and simply ignore the error like you suggested, when we explicitly wait for the child and it already exited before.
I fixed it in the most recent version.